lib/velum/salt.rb

```diff
@@ -72,11 +72,25 @@ def self.accept_minion(minion_id: "")
 
     # Removes a minion from the cluster
     def self.remove_minion(minion_id: "")
-      res = perform_request(endpoint: "/", method: "post",
-                            data: { client: "wheel",
-                                    fun:    "key.delete",
-                                    match:  minion_id })
-      JSON.parse(res.code)
+      # cloud minions need to be terminated, which will also
+      # remove its salt key from the master
+      res = perform_request(endpoint: "/", method:   "post",
+                            data:     { client: "local",
+                                        tgt:    "admin",
+                                        fun:    "cloud.query" })
+      if res.body.include?(minion_id)
+        res = perform_request(endpoint: "/", method:   "post",
+                              data:     { client: "local_async",
+                                          tgt:    "admin",
+                                          fun:    "cloud.destroy",
+                                          arg:    [ minion_id ] })
+      else
+        res = perform_request(endpoint: "/", method: "post",
+                              data: { client: "wheel",
+                                      fun:    "key.delete",
+                                      match:  minion_id })
+      end
+      JSON.parse(res.body)
```

Comment:
This changed from parsing the `res.code` to parsing the whole body of the response. Doesn't it have side effects in other parts of the code?

Reply:
I don't think the return code is used at the moment, and actually even couldn't, as the existing code `JSON.parse(res.code)` throws an exception since `res.code` is not JSON (see also https://bugzilla.suse.com/show_bug.cgi?id=1117152).